src/System/PairExchange.jl

```diff
@@ -14,68 +14,27 @@ function bond_parity(bond)
     return bond_delta > (0, 0, 0, 0)
 end
 
-
-"""
-    set_biquadratic!(sys::System, J, bond::Bond)
-
-Sets a scalar biquadratic interaction along `bond`, yielding a pairwise energy
-``J (𝐒_i⋅𝐒_j)²``. This interaction will be propagated to equivalent bonds in
-consistency with crystal symmetry. Any previous biquadratic exchange
-interactions on these bonds will be overwritten.
-
-For systems restricted to dipoles, the biquadratic interactions will
-automatically be renormalized to achieve maximum consistency with the more
-variationally accurate SU(_N_) mode. This renormalization introduces also a
-correction to the quadratic part of the exchange.
-
-See also [`set_exchange!`](@ref).
-"""
-function set_biquadratic!(sys::System{N}, J, bond::Bond) where N
-    is_homogeneous(sys) || error("Use `set_biquadratic_at!` for an inhomogeneous system.")
-
-    # If `sys` has been reshaped, then operate first on `sys.origin`, which
-    # contains full symmetry information.
-    if !isnothing(sys.origin)
-        set_biquadratic!(sys.origin, J, bond)
-        set_interactions_from_origin!(sys)
-        return
-    end
-
-    validate_bond(sys.crystal, bond)
-
-    ints = interactions_homog(sys)
-
-
-    # Print a warning if an interaction already exists for bond
-    if any(x -> x.bond == bond, ints[bond.i].biquad)
-        println("Warning: Overriding biquadratic interaction for bond $bond.")
-    end
-
-    for i in 1:natoms(sys.crystal)
-        bonds = all_symmetry_related_bonds_for_atom(sys.crystal, i, bond)
-        isempty(bonds) && continue
-
-        for bond′ in bonds
-            # Remove any existing exchange for bond′
-            matches_bond(c) = c.bond == bond′
-            filter!(!matches_bond, ints[i].biquad)
-
-            # The energy or force calculation only needs to see each bond once
-            isculled = bond_parity(bond′)
-
-            # Add to list
-            coupling = Coupling(isculled, bond′, J)
-            push!(ints[i].biquad, coupling)
-        end
-
-        # Sort interactions so that non-culled bonds appear first
-        sort!(ints[i].biquad, by=c->c.isculled)
+# Convert J to Union{Float64, Mat3}
+function to_float_or_mat3(J)
+    if J isa Number || isapprox(J, J[1] * I, atol=1e-12)
```

Comment:
Why is this not a floating point equality check? Is there some situation where we want to round a user's matrix?

Reply:
No, I mean why `isapprox` instead of `==`. I see now that this behavior was already here before the PR, but why?

Reply:
Let's avoid floating point equality checks wherever possible.

As a concrete example, suppose we are generating `J` by a rotation: `J = R * J0 * R'`, where `J0` happens to be exactly diagonal. Mathematically, we should have `R * R' = I`, but numerically this might only be approximate. The approximate equality check will correctly infer that `J` is _intended_ to be diagonal.

Reply:
Also, from the users perspective, the stakes are:

- change in numerical result at the level of 1e-12, presumably not a big deal in most contexts.
- possibly significant speedup

Reply:
After thinking about it more, maybe we should log a warning if this situation is detected.

Reply:
Hmm, I think I agree with you now 😆. The situation is rare, performance difference will likely be small, and using a full Mat3 would never be wrong.

Reply:
I actually quite like the warning idea! Behavior is to use whatever they give us, but if it's numerically close to diagonal, a warning/suggestion that they could make it exactly diagonal for better performance may be warranted!